Sources/Update/update().swift

```diff
@@ -9,18 +9,62 @@
 */
 
 import struct PackageDescription.Version
+import Utility
+import POSIX
+import Get
 
+enum Error: ErrorProtocol {
+    case NoRepo(String)
+    case NoVersion(String)
+    case NoPackageName(String)
+}
 
-public protocol Package {
-    var path: String { get }
-    var version: Version { get }
-    var url: String { get }
+public struct Delta<T> {
+    public var added: [T] = []
+    public var removed: [T] = []
+    public var upgraded: [T] = []
+    public var downgraded: [T] = []
+}
 
-    func fetch()
-    func commit(newVersion: Version)
+public func update(root: String) throws {
+    for path in walk(root, recursively: false) {
+        guard path.isDirectory else { continue }
+        guard let repo = Git.Repo(path: path) else { throw Error.NoRepo(path) }
+        let oldVersion = repo.version
+        try repo.fetch()
+        try repo.upgrade()
+        print("updated:", try repo.pkgname(), oldVersion, "->", repo.version)
+    }
 }
 
+extension Git.Repo {
+    func upgrade() throws {
+        guard let latestVersion = versions.last else { throw Error.NoVersion(path) }
+        let vstr = (versionsArePrefixed ? "v" : "") + latestVersion.description
+        let name = try pkgname()
+        try Git.runPopen([Git.tool, "-C", path, "reset", "--hard", "refs/tags/\(vstr)"])
+        try Git.runPopen([Git.tool, "-C", path, "branch", "-m", vstr])
+        try rename(old: path, new: "\(name)-\(vstr)")
+    }
+
+    /**
+     TODO should be at a higher module level.
+     FIXME DRY
+     FIXME also this is gross and flakier than necessary.
+    */
+    func pkgname() throws -> String {
+        let version = self.version.description.characters
+        let rawname = path.basename.characters.dropLast(version.count + 1)
```

Comment:
the v-prefix is not part of the basename of the path we clone to, so this code is ok.